added file wfcGen.py (main file)

mineserver228 opened the pull request.

mineserver228 committed:
Uploaded main file

mineserver228 committed:
small changes and bug fixes

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+from PIL import Image
+import random
+
```
Check out imoprt formatting suggestions from PEP8.

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+import random
+
+res = 0
+WM = False # 0 - no, 1 - yes, WithMove 
```
Outdated comment. 

Also, a questionable variable name...

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+cropped = []
+for ins in range(4):
+    cropped.append([])
+    if WM == 1:
+        field = Image.open("image.png")
+        
+        SX = field.size[0]
+        SY = field.size[1]
+
+        copy = Image.new("RGB", (SX*2, SY*2))
+        copy.paste(field, (0, 0))
+        copy.paste(field, (SX, 0))
+        copy.paste(field, (0, SY))
+        copy.paste(field, (SX, SY))
+        field = copy
+
+        for y in range(SY):
+            cropped[ins].append([])
+            for x in range(SX):
+                cropped[ins][y].append(field.crop((x, y, x+res, y+res)))
+    else:
+        field = Image.open("image.png")
+        field = field.rotate(90*ins)
+        SX = field.size[0]//res*res
+        SY = field.size[1]//res*res
+        field = field.crop((0, 0, SX, SY))
+
+        for y in range(0, SY, res):
+            cropped[ins].append([])
+            for x in range(0, SY, res):
+                cropped[ins][y//res].append(field.crop((x, y, x+res, y+res)))
```
There is a significant amount of duplicated code here. From trivial things, such as `field = Image.open("image.png")`, which can be just moved out, to a little more complex statements, such as the last pair of for loops, which, I belive, you can slightly rewrite to reduce code duplication

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+                    flag = key
+            
+            if flag != -1:
+                if WM:
+                    cells[flag]["nbg"]["0"].append(cropped[ins][y][x-1-(res-1)])
+                    cells[flag]["nbg"]["1"].append(cropped[ins][y-1-(res-1)][x])
+                    cells[flag]["nbg"]["2"].append(cropped[ins][y][(x+1+(res-1))%lcx])
+                    cells[flag]["nbg"]["3"].append(cropped[ins][(y+1+(res-1))%lcy][x])
+                else:
+                    cells[flag]["nbg"]["0"].append(cropped[ins][y][x-1])
+                    cells[flag]["nbg"]["1"].append(cropped[ins][y-1][x])
+                    cells[flag]["nbg"]["2"].append(cropped[ins][y][(x+1)%lcx])
+                    cells[flag]["nbg"]["3"].append(cropped[ins][(y+1)%lcy][x])
+            else:
+                if WM:
+                    cells[str(len(cells.keys()))] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"]["0"] = [cropped[ins][y][x-1-(res-1)]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["1"] = [cropped[ins][y-1-(res-1)][x]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["2"] = [cropped[ins][y][(x+1+(res-1))%lcx]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["3"] = [cropped[ins][(y+1+(res-1))%lcy][x]]
+                    cells[str(len(cells.keys())-1)]["img"] = cropped[ins][y][x]
+                else:
+                    cells[str(len(cells.keys()))] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"]["0"] = [cropped[ins][y][x-1]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["1"] = [cropped[ins][y-1][x]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["2"] = [cropped[ins][y][(x+1)%lcx]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["3"] = [cropped[ins][(y+1)%lcy][x]]
+                    cells[str(len(cells.keys())-1)]["img"] = cropped[ins][y][x]
```
Even more code duplication here, which is probably even easier to fix

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+                    cells[str(len(cells.keys())-1)]["nbg"]["3"] = [cropped[ins][(y+1)%lcy][x]]
+                    cells[str(len(cells.keys())-1)]["img"] = cropped[ins][y][x]
+
+    for y in range(lcy):
+        for x in range(lcx):
+            for key in cells.keys():
+                if cells[key]["img"] == cropped[ins][y][x]:
+                    Gkey = key
+            for key in cells.keys():
+                for i in range(len(cells[key]["nbg"]["0"])):
+                    if cells[key]["nbg"]["0"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["0"][i] = Gkey
+                for i in range(len(cells[key]["nbg"]["1"])):
+                    if cells[key]["nbg"]["1"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["1"][i] = Gkey
+                for i in range(len(cells[key]["nbg"]["2"])):
+                    if cells[key]["nbg"]["2"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["2"][i] = Gkey
+                for i in range(len(cells[key]["nbg"]["3"])):
+                    if cells[key]["nbg"]["3"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["3"][i] = Gkey
```
Why is this done separately from the previous block of for loops? Is there a technical reason? Is it for the sake of code readability?

mineserver228 replied:
yes it is for technical reason

mineserver228 replied:
and i won't fix it

kolayne replied:
What kind of reason? Why will you not fix it?

mineserver228 replied:
because firstly programm adds images to the dictionary and links to other images as neighbours
and secondly when the dictionary is full, it changes llinks to the images on other tile ids
if you will do this at the same time, programm may try to change image on id that is not real

kolayne replied:
Alright. In that case, I wonder, if you can rewrite this code such that instead of having four different for loops (for each of the `cells[key]["nbg"]["0"]`, `["1"]`, `["2"]`, `["3"]`) you have a general for loop through all keys/values of `cells[key]["nbg"]`.

By the way, I can see you are doing something like
```python
for i in range(len(some_list)):
    do_something_with(some_list[i])
```
So you are iterating through indices of `some_list` but never using the index directly, only the corresponding element. In such cases it is better to write
```python
for elem in some_list:
    do_something_with(elem)
```

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+        self.anyFlag = True
+    
+    def collapse(self):
+        '''
+        COLAPS YEAHHH
+        '''
```
???

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+def intersection(mass1, mass2):
+    return list(set(mass1) & set(mass2))
+
+class Cell:
```
It is hard to read this code because you first have something in the procedural style (for loops on the most global level), and then a class definition. It's better to define classes as functions first. And, for that matter, not have any global logic at all (keep everything in functions and classes). But we can leave the latter for a future pull request

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+    def loadEntropy(self, grid):
+        if not self.collapsed:
+            self.anyFlag = True
+            if grid[self.gp[1]-1][self.gp[0]].collapsed:
+                self.entropy = grid[self.gp[1]-1][self.gp[0]].nbg["3"]
+                self.anyFlag = False
+            elif grid[self.gp[1]][self.gp[0]-1].collapsed:
+                self.entropy = grid[self.gp[1]][self.gp[0]-1].nbg["2"]
+                self.anyFlag = False
+            elif (self.gp[1]+1)<len(grid):
+                if grid[self.gp[1]+1][self.gp[0]].collapsed:
+                    self.entropy = grid[self.gp[1]+1][self.gp[0]].nbg["1"]
+                    self.anyFlag = False
+                elif (self.gp[0]+1)<len(grid[0]):
+                    if grid[self.gp[1]][self.gp[0]+1].collapsed:
+                        self.entropy = grid[self.gp[1]][self.gp[0]+1].nbg["0"]
+                        self.anyFlag = False
+                    else:
+                        return
```
Again, hard to read, a lot of duplication. Please, fix here as well as in other places

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+        self.gp = gp
+        self.collapsed = False
+        self.entropy = []
+        self.anyFlag = True
```
What the hell is "anyFlag"?

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+            for x in range(0, self.sx):
+                if not self.grid[y][x].collapsed:
+                    self.grid[y][x].loadEntropy(self.grid)
+                    if self.grid[y][x].getEntropy() in self.entropySave.keys():
+                        self.entropySave[self.grid[y][x].getEntropy()].append([x, y])
+                    else:
+                        self.entropySave[self.grid[y][x].getEntropy()] = [[x, y]]
```
Firstly, notice how you're calling `self.grid[y][x].getEntropy()` several times. Instead, you could save it into a variable.

Secondly, I would suggest that you use the `dict.get` method with a second parameter here.

```suggestion
                    entropy = self.grid[y][x].getEntropy()
                    self.entropySave[entropy] = self.entropySave.get(entropy, []) + [[x, y]]
```

By the way, it is common to store coordinates as tuples (`(x, y)`) rather than lists (`[x, y]`)

mineserver228 replied:
i'm saving them like that in the case if I would need to change them

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+    def collapse(self):
+        try:
+            self.loadEntropy()
+            s = sorted(self.entropySave.keys())
```
Bad variable name

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+            
+            self.grid[y][x].collapse()
+            self.loadEntropy()
+            flag1 = 0 in self.entropySave.keys()
```
Bad variable name

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+            if ITERATION % 400 == 0:
+                self.getFinalImage()
+            return False
+        except:
```
Too broad `except` clause. Consider specifying a specific exception that you mean to catch

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,287 @@
+        
+    
+    def getFinalImage(self):
+        if self.GIMAGE == None:
```
In python the `is` operator is commonly used here:

```suggestion
        if self.GIMAGE is None:
```

Technical reason: https://stackoverflow.com/questions/14247373/python-none-comparison-should-i-use-is-or

I don't think it often happens that it makes a difference but... ¯\_(ツ)_/¯

kolayne reviewed (changes requested):
The main problems of the code at the moment IMO is lots of code duplication. Consider improving it!

mineserver228 committed:
Bug fix and small code changes

- fixed bug when in crop shift mode programm was not rotating full image
  four times
- changed import formatting to PEP8
- deleted most of duplicated code
- changed names of some variables
  WM -> cropShift
  flag1 -> ZeroCheck
- deleted unnecessary code and comments

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,264 @@
+
+        field = field.rotate(90*ins)
+
+        SX = field.size[0]
+        SY = field.size[1]
```
This can be expressed as
```suggestion
        SX, SY = field.size
```

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,264 @@
+        copy.paste(field, (SX, 0))
+        copy.paste(field, (0, SY))
+        copy.paste(field, (SX, SY))
+        field = copy
```
Why?
As python objects are reference types, `field` and `copy` become the same image. Not copies of the same image (which would happen if you did `field = copy.copy()`) but the exact same image. When you modify an object through `field`, the object visible through `copy` is modified the same way (because they point at the same object).

You, most likely, either mean to copy the image (via the `.copy()` method) or can just use one name for the image

mineserver228 replied:
I would never use copy again, so i don't think that it would break anything

kolayne replied:
Then why do you have it in the first place?

It is not very safe to assume you will never use it again. When you come back to this code 1 year later, will you still remember that these two variables are pointing at the same object and you're not supposed to use copy?

You could make your language technically ensure that you don't (e.g., in Python you could write `del copy`) but in this case - why would you? I think it makes more sense to just create the image and call it `field`

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,264 @@
+cells = {}
+cropped = []
+for ins in range(4):
+    field = Image.open("image.png")
+    if cropShift:
+
+        field = field.rotate(90*ins)
```
These line breaks are rather confusing IMO. Do you think the line `if cropShift:` is logically closer to `field = Image.open("image.png")` than to `field = field.rotate(90*ins)`?

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,264 @@
+                cells[flag]["nbg"]["2"].append(cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx])
+                cells[flag]["nbg"]["3"].append(cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x])
+            else:
+                cells[str(len(cells.keys()))] = {}
+                cells[str(len(cells.keys())-1)]["nbg"] = {}
+                cells[str(len(cells.keys())-1)]["nbg"]["0"] = [cropped[ins][y][x-1-(res-1)*int(cropShift)]]
+                cells[str(len(cells.keys())-1)]["nbg"]["1"] = [cropped[ins][y-1-(res-1)*int(cropShift)][x]]
+                cells[str(len(cells.keys())-1)]["nbg"]["2"] = [cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx]]
+                cells[str(len(cells.keys())-1)]["nbg"]["3"] = [cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x]]
+                cells[str(len(cells.keys())-1)]["img"] = cropped[ins][y][x]
```
1. Why `str(len(cells.keys()))` rather than just `len(cells.keys())`?
2. Why do you create an empty dictionary and then make four additions to it? You can create the dictionary with these values already there. If you care about the nice multiline formatting (which I _would_ care about, if I were you), there is no issue with that as well: you can write a multi-line declaration of a dictionary, such as
```python
d = {
    1: "aba",
    2: "caba",
    3: "abacaba",
}
```

kolayne reviewed (changes requested):
A few new comments. Besides, there are some old ones not yet fixed

mineserver228 committed:
critical bug fixes and some code changes

- fixed two critical bugs
- shortend loadEntropy function and some other to make them more
  readable
- renamed variable
  anyFlag -> checkCollapsedNeighbours
- deleted more unnecessary code

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,244 @@
+                cells[flag]["nbg"]["2"].append(cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx])
+                cells[flag]["nbg"]["3"].append(cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x])
+            else:
+                cells[str(len(cells.keys()))] = {
+                "nbg" : {
+                    "0" : [cropped[ins][y][x-1-(res-1)*int(cropShift)]],
+                    "1" : [cropped[ins][y-1-(res-1)*int(cropShift)][x]],
+                    "2" : [cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx]],
+                    "3" : [cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x]],
+                }, 
+                "img" : cropped[ins][y][x]}
```
Formatting is not very easy to read here:
```suggestion
                cells[str(len(cells.keys()))] = {
                    "nbg" : {
                        "0" : [cropped[ins][y][x-1-(res-1)*int(cropShift)]],
                        "1" : [cropped[ins][y-1-(res-1)*int(cropShift)][x]],
                        "2" : [cropped[ins][y][(x+1+(res-1)*int(cropShift))%lcx]],
                        "3" : [cropped[ins][(y+1+(res-1)*int(cropShift))%lcy][x]],
                    }, 
                    "img" : cropped[ins][y][x]
                }
```

kolayne commented on wfcGen.py:
```diff
@@ -0,0 +1,244 @@
+        if not self.collapsed:
+            self.checkCollapsedNeighbours = True
+
+            for xy in range(-1, 1, 2):
```
```suggestion
            for xy in [-1, 1]:
```

mineserver228 committed:
keys in cells dictionary changed from str to int type

mineserver228 committed:
small code changes to make it more readable
